File: src/main/java/maestro/pathing/calc/Path.java
package maestro.pathing.calc;

import com.google.common.collect.Lists;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import maestro.api.pathing.calc.IPath;
import maestro.api.pathing.goals.Goal;
import maestro.api.pathing.movement.IMovement;
import maestro.api.utils.BetterBlockPos;
import maestro.api.utils.Helper;
import maestro.pathing.movement.CalculationContext;
import maestro.pathing.movement.Movement;
import maestro.pathing.movement.Moves;
import maestro.pathing.path.CutoffPath;
import maestro.utils.pathing.PathBase;

/**
 * A node based implementation of IPath
 *
 * @author leijurv
 */
class Path extends PathBase {

    /** The start position of this path */
    private final BetterBlockPos start;

    /** The end position of this path */
    private final BetterBlockPos end;

    /**
     * The blocks on the path. Guaranteed that path.get(0) equals start and path.get(path.size()-1)
     * equals end
     */
    private final List<BetterBlockPos> path;

    private final List<Movement> movements;

    private final List<PathNode> nodes;

    private final Goal goal;

    private final int numNodes;

    private final CalculationContext context;

    private volatile boolean verified;

    Path(
            BetterBlockPos realStart,
            PathNode start,
            PathNode end,
            int numNodes,
            Goal goal,
            CalculationContext context) {
        this.end = new BetterBlockPos(end.x, end.y, end.z);
        this.numNodes = numNodes;
        this.movements = new ArrayList<>();
        this.goal = goal;
        this.context = context;

        PathNode current = end;
        List<BetterBlockPos> tempPath = new ArrayList<>();
        List<PathNode> tempNodes = new ArrayList<>();
        while (current != null) {
            tempNodes.add(current);
            tempPath.add(new BetterBlockPos(current.x, current.y, current.z));
            current = current.previous;
        }

        // If the position the player is at is different from the position we told A* to start from,
        // and A* gave us no movements, then add a fake node that will allow a movement to be
        // created
        // that gets us to the single position in the path.
        // See PathingBehavior#createPathfinder and https://github.com/cabaletta/baritone/pull/4519
        var startNodePos = new BetterBlockPos(start.x, start.y, start.z);
        if (!realStart.equals(startNodePos) && start.equals(end)) {
            this.start = realStart;
            PathNode fakeNode = new PathNode(realStart.x, realStart.y, realStart.z, goal);
            fakeNode.cost = 0;
            tempNodes.add(fakeNode);
            tempPath.add(realStart);
        } else {
            this.start = startNodePos;
        }

        // Nodes are traversed last to first so we need to reverse the list
        this.path = Lists.reverse(tempPath);
        this.nodes = Lists.reverse(tempNodes);
    }

    @Override
    public Goal getGoal() {
        return goal;
    }

    private boolean assembleMovements() {
        if (path.isEmpty() || !movements.isEmpty()) {
            throw new IllegalStateException("Path must not be empty");
        }
        for (int i = 0; i < path.size() - 1; i++) {
            double cost = nodes.get(i + 1).cost - nodes.get(i).cost;
            PathNode nextNode = nodes.get(i + 1);

            // NEW: Check if movement was stored directly in node
            if (nextNode.previousMovement != null) {
                Movement move = (Movement) nextNode.previousMovement;
                // Verify destination matches (sanity check)
                if (move.getDest().equals(path.get(i + 1))) {
                    movements.add(move);
                    continue;
                } else {
                    // Shouldn't happen, but log and fallback
                    Helper.HELPER.logDebug(
                            "Stored movement destination mismatch: expected "
                                    + path.get(i + 1)
                                    + ", got "
                                    + move.getDest());
                }
            }

            // FALLBACK: Use old runBackwards method for compatibility
            Movement move = runBackwards(path.get(i), path.get(i + 1), nextNode, cost);
            if (move == null) {
                return true;
            } else {
                movements.add(move);
            }
        }
        return false;
    }

    private Movement runBackwards(
            BetterBlockPos src, BetterBlockPos dest, PathNode destNode, double cost) {
        // Try recorded movement first (O(1) lookup)
        Moves recordedMove = destNode.getMovement();
        if (recordedMove != null) {
            Movement move = recordedMove.apply0(context, src);
            if (move.getDest().equals(dest)) {
                // have to calculate the cost at calculation time so we can accurately judge whether
                // a cost increase happened between cached calculation and real execution
                // however, taking into account possible favoring that could skew the node cost, we
                // really want the stricter limit of the two
                // so we take the minimum of the path node cost difference, and the calculated cost
                move.override(Math.min(move.calculateCost(context), cost));
                return move;
            }
            // Recorded movement doesn't match destination (shouldn't happen)
            Helper.HELPER.logDebug(
                    "Recorded movement "
                            + recordedMove
                            + " from "
                            + src
                            + " doesn't match expected dest "
                            + dest);
        }

        // Fallback: Linear search through all movements (for backward compatibility)
        for (Moves moves : Moves.values()) {
            Movement move = moves.apply0(context, src);
            if (move.getDest().equals(dest)) {
                move.override(Math.min(move.calculateCost(context), cost));
                return move;
            }
        }

        // this is no longer called from bestPathSoFar, now it's in postprocessing
        Helper.HELPER.logDebug(
                "Movement became impossible during calculation "
                        + src
                        + " "
                        + dest
                        + " "
                        + dest.subtract(src));
        return null;
    }

    @Override
    public IPath postProcess() {
        if (verified) {
            throw new IllegalStateException("Path must not be verified twice");
        }
        verified = true;
        boolean failed = assembleMovements();
        movements.forEach(m -> m.checkLoadedChunk(context));

        if (failed) { // at least one movement became impossible during calculation
            CutoffPath res = new CutoffPath(this, movements().size());
            if (res.movements().size() != movements.size()) {
                throw new IllegalStateException("Path has wrong size after cutoff");
            }
            return res;
        }
        // more post processing here
        sanityCheck();
        return this;
    }

    @Override
    public List<IMovement> movements() {
        if (!verified) {
            // edge case note: this is called during verification
            throw new IllegalStateException("Path not yet verified");
        }
        return Collections.unmodifiableList(movements);
    }

    @Override
    public List<BetterBlockPos> positions() {
        return Collections.unmodifiableList(path);
    }

    @Override
    public int getNumNodesConsidered() {
        return numNodes;
    }

    @Override
    public BetterBlockPos getSrc() {
        return start;
    }

    @Override
    public BetterBlockPos getDest() {
        return end;
    }
}

File: common/src/main/maestro/pathing/calc/Path.kt
package maestro.pathing.calc

import maestro.api.pathing.calc.IPath
import maestro.api.pathing.goals.Goal
import maestro.api.pathing.movement.IMovement
import maestro.api.utils.MaestroLogger
import maestro.api.utils.PackedBlockPos
import maestro.pathing.PathBase
import maestro.pathing.movement.CalculationContext
import maestro.pathing.movement.Movement
import maestro.pathing.path.CutoffPath
import org.slf4j.Logger

/** A node based implementation of IPath */
internal class Path(
    realStart: PackedBlockPos,
    start: PathNode,
    end: PathNode,
    private val numNodes: Int,
    goal_: Goal,
    private val context: CalculationContext,
) : PathBase() {
    override val goal: Goal = goal_
    override val numNodesConsidered: Int get() = numNodes

    /** The start position of this path */
    private val start: PackedBlockPos

    /** The end position of this path */
    private val end: PackedBlockPos = PackedBlockPos(end.x, end.y, end.z)

    /**
     * The blocks on the path. Guaranteed that path.get(0) equals start and path.get(path.size()-1)
     * equals end
     */
    private val path: List<PackedBlockPos>

    private val movements: MutableList<Movement> = mutableListOf()

    private val nodes: List<PathNode>

    @Volatile
    private var verified: Boolean = false

    init {
        // Build path by walking backwards from end to start
        var current: PathNode? = end
        val tempPath = mutableListOf<PackedBlockPos>()
        val tempNodes = mutableListOf<PathNode>()

        while (current != null) {
            tempNodes.add(current)
            tempPath.add(PackedBlockPos(current.x, current.y, current.z))
            current = current.previous
        }

        // If the position the player is at is different from the position we told A* to start from,
        // and A* gave us no movements, then add a fake node that will allow a movement to be
        // created that gets us to the single position in the path.
        // See PathingBehavior#createPathfinder and https://github.com/cabaletta/baritone/pull/4519
        val startNodePos = PackedBlockPos(start.x, start.y, start.z)
        if (realStart != startNodePos && start == end) {
            this.start = realStart
            val fakeNode =
                PathNode(realStart.x, realStart.y, realStart.z, goal).apply {
                    cost = 0.0
                }
            tempNodes.add(fakeNode)
            tempPath.add(realStart)
        } else {
            this.start = startNodePos
        }

        // Nodes are traversed last to first so we need to reverse the list
        path = tempPath.reversed()
        nodes = tempNodes.reversed()
    }

    private fun assembleMovements(): Boolean {
        if (path.isEmpty() || movements.isNotEmpty()) {
            throw IllegalStateException("Path must not be empty")
        }

        for (i in 0 until path.size - 1) {
            val nextNode = nodes[i + 1]

            // Get movement from node - all paths should have previousMovement set
            val prevMovement = nextNode.previousMovement
            if (prevMovement == null) {
                // Movement not recorded - path became impossible
                log
                    .atDebug()
                    .addKeyValue("source", path[i])
                    .addKeyValue("dest", path[i + 1])
                    .log("Movement not recorded - path became impossible during calculation")
                return true
            }

            val move = prevMovement as Movement
            // Verify destination matches (sanity check)
            if (move.dest == path[i + 1]) {
                movements.add(move)
            } else {
                // Shouldn't happen - log and fail
                log
                    .atDebug()
                    .addKeyValue("expected_dest", path[i + 1])
                    .addKeyValue("actual_dest", move.dest)
                    .log("Stored movement destination mismatch - path became impossible")
                return true
            }
        }
        return false
    }

    override fun postProcess(): IPath {
        if (verified) {
            throw IllegalStateException("Path must not be verified twice")
        }
        verified = true
        val failed = assembleMovements()
        movements.forEach { it.checkLoadedChunk(context) }

        if (failed) { // at least one movement became impossible during calculation
            val res = CutoffPath(this, movements().size)
            if (res.movements().size != movements.size) {
                throw IllegalStateException("Path has wrong size after cutoff")
            }
            return res
        }
        // more post-processing here
        sanityCheck()
        return this
    }

    override fun movements(): List<IMovement> {
        if (!verified) {
            // edge case note: this is called during verification
            throw IllegalStateException("Path not yet verified")
        }
        return movements.toList()
    }

    override fun positions(): List<PackedBlockPos> = path

    override fun replaceMovement(
        index: Int,
        newMovement: IMovement,
    ) {
        if (!verified) {
            throw IllegalStateException("Cannot replace movement in unverified path")
        }
        if (index < 0 || index >= movements.size) {
            throw IndexOutOfBoundsException(
                "Index $index out of bounds for movements list of size ${movements.size}",
            )
        }
        movements[index] = newMovement as Movement
    }

    companion object {
        private val log: Logger = MaestroLogger.get("path")
    }
}
